File: jung-io/src/main/java/edu/uci/ics/jung/io/graphml/parser/GraphElementParser.java
/*
 * Copyright (c) 2008, The JUNG Authors
 *
 * All rights reserved.
 *
 * This software is open-source under the BSD license; see either
 * "license.txt" or
 * https://github.com/jrtom/jung/blob/master/LICENSE for a description.
 */

package edu.uci.ics.jung.io.graphml.parser;

import com.google.common.graph.MutableNetwork;
import edu.uci.ics.jung.io.GraphIOException;
import edu.uci.ics.jung.io.graphml.DataMetadata;
import edu.uci.ics.jung.io.graphml.EdgeMetadata;
import edu.uci.ics.jung.io.graphml.ExceptionConverter;
import edu.uci.ics.jung.io.graphml.GraphMLConstants;
import edu.uci.ics.jung.io.graphml.GraphMetadata;
import edu.uci.ics.jung.io.graphml.GraphMetadata.EdgeDefault;
import edu.uci.ics.jung.io.graphml.NodeMetadata;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.Map;
import javax.xml.stream.XMLEventReader;
import javax.xml.stream.events.Attribute;
import javax.xml.stream.events.EndElement;
import javax.xml.stream.events.StartElement;
import javax.xml.stream.events.XMLEvent;

/**
 * Parses graph elements.
 *
 * @author Nathan Mittler - nathan.mittler@gmail.com
 */
public class GraphElementParser<G extends MutableNetwork<N, E>, N, E>
    extends AbstractElementParser<G, N, E> {

  public GraphElementParser(ParserContext<G, N, E> parserContext) {
    super(parserContext);
  }

  public GraphMetadata parse(XMLEventReader xmlEventReader, StartElement start)
      throws GraphIOException {

    try {
      // Create the new graph.
      GraphMetadata graphMetadata = new GraphMetadata();

      // Parse the attributes.
      @SuppressWarnings("unchecked")
      Iterator<Attribute> iterator = start.getAttributes();
      while (iterator.hasNext()) {
        Attribute attribute = iterator.next();
        String name = attribute.getName().getLocalPart();
        String value = attribute.getValue();
        if (graphMetadata.getId() == null && GraphMLConstants.ID_NAME.equals(name)) {

          graphMetadata.setId(value);
        } else if (graphMetadata.getEdgeDefault() == null
            && GraphMLConstants.EDGEDEFAULT_NAME.equals(name)) {

          graphMetadata.setEdgeDefault(
              GraphMLConstants.DIRECTED_NAME.equals(value)
                  ? EdgeDefault.DIRECTED
                  : EdgeDefault.UNDIRECTED);
        } else {
          graphMetadata.setProperty(name, value);
        }
      }

      // Make sure the graphdefault has been set.
      if (graphMetadata.getEdgeDefault() == null) {
        throw new GraphIOException("Element 'graph' is missing attribute 'edgedefault'");
      }

      Map<String, N> idToNodeMap = new HashMap<String, N>();
      Collection<EdgeMetadata> edgeMetadata = new LinkedList<EdgeMetadata>();
      //            Collection<HyperEdgeMetadata> hyperEdgeMetadata = new
      // LinkedList<HyperEdgeMetadata>();

      while (xmlEventReader.hasNext()) {

        XMLEvent event = xmlEventReader.nextEvent();
        if (event.isStartElement()) {
          StartElement element = (StartElement) event;

          String name = element.getName().getLocalPart();
          if (GraphMLConstants.DESC_NAME.equals(name)) {

            // Parse the description and set it in the graph.
            String desc = (String) getParser(name).parse(xmlEventReader, element);
            graphMetadata.setDescription(desc);

          } else if (GraphMLConstants.DATA_NAME.equals(name)) {

            // Parse the data element and store the property in the graph.
            DataMetadata data = (DataMetadata) getParser(name).parse(xmlEventReader, element);
            graphMetadata.addData(data);

          } else if (GraphMLConstants.NODE_NAME.equals(name)) {

            // Parse the node metadata
            NodeMetadata metadata = (NodeMetadata) getParser(name).parse(xmlEventReader, element);

            // Create the node object and store it in the metadata
            N node = getParserContext().createNode(metadata);
            metadata.setNode(node);
            idToNodeMap.put(metadata.getId(), node);

            // Add it to the graph
            graphMetadata.addNodeMetadata(node, metadata);

          } else if (GraphMLConstants.EDGE_NAME.equals(name)) {

            // Parse the edge metadata
            EdgeMetadata metadata = (EdgeMetadata) getParser(name).parse(xmlEventReader, element);

            // Set the directed property if not overridden.
            if (metadata.isDirected() == null) {
              metadata.setDirected(graphMetadata.getEdgeDefault() == EdgeDefault.DIRECTED);
            }

            // Create the edge object and store it in the metadata
            E edge = getParserContext().createEdge(metadata);
            edgeMetadata.add(metadata);
            metadata.setEdge(edge);

            // Add it to the graph.
            graphMetadata.addEdgeMetadata(edge, metadata);

            //                    } else if (GraphMLConstants.HYPEREDGE_NAME.equals(name)) {
            //
            //                        // Parse the edge metadata
            //                        HyperEdgeMetadata metadata = (HyperEdgeMetadata)
            // getParser(name).parse(
            //                                xmlEventReader, element);
            //
            //                        // Create the edge object and store it in the metadata
            //                        E edge = getParserContext().createHyperEdge(metadata);
            //                        hyperEdgeMetadata.add(metadata);
            //                        metadata.setEdge(edge);
            //
            //                        // Add it to the graph
            //                        graphMetadata.addHyperEdgeMetadata(edge, metadata);

          } else {

            // Treat anything else as unknown
            getUnknownParser().parse(xmlEventReader, element);
          }
        }
        if (event.isEndElement()) {
          EndElement end = (EndElement) event;
          verifyMatch(start, end);
          break;
        }
      }

      // Apply the keys to this object.
      applyKeys(graphMetadata);

      // Create the graph object and store it in the metadata
      G graph = getParserContext().createGraph(graphMetadata);
      graphMetadata.setGraph(graph);

      // Add all of the nodes to the graph object.
      addNodesToGraph(graph, idToNodeMap.values());

      // Add the edges to the graph object.
      addEdgesToGraph(graph, edgeMetadata, idToNodeMap);
      //            addHyperEdgesToGraph(graph, hyperEdgeMetadata, idToNodeMap);

      return graphMetadata;

    } catch (Exception e) {
      ExceptionConverter.convert(e);
    }

    return null;
  }

  private void addNodesToGraph(G graph, Collection<N> nodes) {

    for (N node : nodes) {
      graph.addNode(node);
    }
  }

  @SuppressWarnings("unchecked")
  private void addEdgesToGraph(
      G graph, Collection<EdgeMetadata> metadata, Map<String, N> idToNodeMap)
      throws GraphIOException {

    for (EdgeMetadata emd : metadata) {

      // Get the edge out of the metadata
      E edge = (E) emd.getEdge();

      // Get the nodes.
      N source = idToNodeMap.get(emd.getSource());
      N target = idToNodeMap.get(emd.getTarget());
      if (source == null || target == null) {
        throw new GraphIOException(
            "edge references undefined source or target node. "
                + "Source: "
                + emd.getSource()
                + ", Target: "
                + emd.getTarget());
      }

      // Add it to the graph.
      graph.addEdge(source, target, edge);
    }
  }

  // TODO: hypergraph support
  //    @SuppressWarnings("unchecked")
  //    private void addHyperEdgesToGraph(G graph, Collection<HyperEdgeMetadata> metadata,
  //            Map<String,V> idToNodeMap) throws GraphIOException {
  //
  //        for (HyperEdgeMetadata emd : metadata) {
  //
  //            // Get the edge out of the metadata
  //            E edge = (E)emd.getEdge();
  //
  //            // Add the verticies to a list.
  //            List<N> verticies = new ArrayList<N>();
  //            List<EndpointMetadata> endpoints = emd.getEndpoints();
  //            for (EndpointMetadata ep : endpoints) {
  //                N v = idToNodeMap.get(ep.getNode());
  //                if (v == null) {
  //                    throw new GraphIOException(
  //                            "hyperedge references undefined node: "
  //                                    + ep.getNode());
  //                }
  //                verticies.add(v);
  //            }
  //
  //            // Add it to the graph.
  //            graph.addEdge(edge, verticies);
  //        }
  //    }
}

File: jung-io/src/main/kotlin/edu/uci/ics/jung/io/graphml/parser/GraphElementParser.kt
/*
 * Copyright (c) 2008, The JUNG Authors
 *
 * All rights reserved.
 *
 * This software is open-source under the BSD license; see either
 * "license.txt" or
 * https://github.com/jrtom/jung/blob/master/LICENSE for a description.
 */
package edu.uci.ics.jung.io.graphml.parser

import com.google.common.graph.MutableNetwork
import edu.uci.ics.jung.io.GraphIOException
import edu.uci.ics.jung.io.graphml.DataMetadata
import edu.uci.ics.jung.io.graphml.EdgeMetadata
import edu.uci.ics.jung.io.graphml.ExceptionConverter
import edu.uci.ics.jung.io.graphml.GraphMLConstants
import edu.uci.ics.jung.io.graphml.GraphMetadata
import edu.uci.ics.jung.io.graphml.GraphMetadata.EdgeDefault
import edu.uci.ics.jung.io.graphml.NodeMetadata
import java.util.LinkedList
import javax.xml.stream.XMLEventReader
import javax.xml.stream.events.Attribute
import javax.xml.stream.events.EndElement
import javax.xml.stream.events.StartElement

/**
 * Parses graph elements.
 *
 * @author Nathan Mittler - nathan.mittler@gmail.com
 */
class GraphElementParser<G : MutableNetwork<N, E>, N, E>(
  parserContext: ParserContext<G, N, E>
) : AbstractElementParser<G, N, E>(parserContext) {

  @Throws(GraphIOException::class)
  override fun parse(xmlEventReader: XMLEventReader, start: StartElement): GraphMetadata? {
    try {
      // Create the new graph.
      val graphMetadata = GraphMetadata()

      // Parse the attributes.
      @Suppress("UNCHECKED_CAST")
      val iterator = start.attributes as Iterator<Attribute>
      while (iterator.hasNext()) {
        val attribute = iterator.next()
        val name = attribute.name.localPart
        val value = attribute.value
        if (graphMetadata.id == null && GraphMLConstants.ID_NAME == name) {
          graphMetadata.id = value
        } else if (graphMetadata.edgeDefault == null &&
          GraphMLConstants.EDGEDEFAULT_NAME == name
        ) {
          graphMetadata.edgeDefault =
            if (GraphMLConstants.DIRECTED_NAME == value) EdgeDefault.DIRECTED
            else EdgeDefault.UNDIRECTED
        } else {
          graphMetadata.setProperty(name, value)
        }
      }

      // Make sure the graphdefault has been set.
      if (graphMetadata.edgeDefault == null) {
        throw GraphIOException("Element 'graph' is missing attribute 'edgedefault'")
      }

      val idToNodeMap = HashMap<String, N>()
      val edgeMetadata: MutableCollection<EdgeMetadata> = LinkedList()

      while (xmlEventReader.hasNext()) {
        val event = xmlEventReader.nextEvent()
        if (event.isStartElement) {
          val element = event as StartElement
          val name = element.name.localPart
          if (GraphMLConstants.DESC_NAME == name) {
            // Parse the description and set it in the graph.
            val desc = getParser(name).parse(xmlEventReader, element) as String
            graphMetadata.description = desc
          } else if (GraphMLConstants.DATA_NAME == name) {
            // Parse the data element and store the property in the graph.
            val data = getParser(name).parse(xmlEventReader, element) as DataMetadata
            graphMetadata.addData(data)
          } else if (GraphMLConstants.NODE_NAME == name) {
            // Parse the node metadata
            val metadata = getParser(name).parse(xmlEventReader, element) as NodeMetadata

            // Create the node object and store it in the metadata
            val node = parserContext.createNode(metadata)
            metadata.node = node
            idToNodeMap[metadata.id!!] = node

            // Add it to the graph
            graphMetadata.addNodeMetadata(node as Any, metadata)
          } else if (GraphMLConstants.EDGE_NAME == name) {
            // Parse the edge metadata
            val metadata = getParser(name).parse(xmlEventReader, element) as EdgeMetadata

            // Set the directed property if not overridden.
            if (metadata.isDirected == null) {
              metadata.isDirected = graphMetadata.edgeDefault == EdgeDefault.DIRECTED
            }

            // Create the edge object and store it in the metadata
            val edge = parserContext.createEdge(metadata)
            edgeMetadata.add(metadata)
            metadata.edge = edge

            // Add it to the graph.
            graphMetadata.addEdgeMetadata(edge as Any, metadata)
          } else {
            // Treat anything else as unknown
            getUnknownParser().parse(xmlEventReader, element)
          }
        }
        if (event.isEndElement) {
          val end = event as EndElement
          verifyMatch(start, end)
          break
        }
      }

      // Apply the keys to this object.
      applyKeys(graphMetadata)

      // Create the graph object and store it in the metadata
      val graph = parserContext.createGraph(graphMetadata)
      graphMetadata.graph = graph

      // Add all of the nodes to the graph object.
      addNodesToGraph(graph, idToNodeMap.values)

      // Add the edges to the graph object.
      addEdgesToGraph(graph, edgeMetadata, idToNodeMap)

      return graphMetadata
    } catch (e: Exception) {
      ExceptionConverter.convert(e)
    }

    return null
  }

  private fun addNodesToGraph(graph: G, nodes: Collection<N>) {
    for (node in nodes) {
      graph.addNode(node)
    }
  }

  @Suppress("UNCHECKED_CAST")
  @Throws(GraphIOException::class)
  private fun addEdgesToGraph(
    graph: G,
    metadata: Collection<EdgeMetadata>,
    idToNodeMap: Map<String, N>
  ) {
    for (emd in metadata) {
      // Get the edge out of the metadata
      val edge = emd.edge as E

      // Get the nodes.
      val source = idToNodeMap[emd.source]
      val target = idToNodeMap[emd.target]
      if (source == null || target == null) {
        throw GraphIOException(
          "edge references undefined source or target node. " +
            "Source: " + emd.source + ", Target: " + emd.target
        )
      }

      // Add it to the graph.
      graph.addEdge(source, target, edge)
    }
  }
}
